File: Sample/Source/Windows/General/ConfigCheckBoxRow.py
# -*- encoding: utf-8 -*-

from __future__ import annotations
from typing import Callable, Optional
from Engine import Image, Vector2f, UI, FloatRect
from Engine.UI import Canvas, PlainText
from Engine.UI.Base import FunctionalBase
from .CheckBox import CheckBox


_ROW_HEIGHT: int = 32
_LABEL_X: int = 32
_LABEL_FONT_SIZE: int = 20


class ConfigCheckBoxRow(Canvas, FunctionalBase):
    r"""\brief Single configuration row combining a label and a CheckBox."""

    def __init__(
        self,
        labelText: str,
        rowWidth: int,
        checkBoxSize: int,
        windowSkin: Optional[Image],
        checked: bool = False,
        onCheckedChanged: Optional[Callable[[bool], None]] = None,
    ) -> None:
        r"""\brief Construct a configuration checkbox row.

        - \param labelText         Localized label shown on the left
        - \param rowWidth          Total row width in logical UI units
        - \param checkBoxSize      Checkbox width and height
        - \param windowSkin        Windowskin shared with the parent window
        - \param checked           Initial checked state
        - \param onCheckedChanged  Callback invoked when the checked state changes
        """
        Canvas.__init__(self, ((0, 0), (rowWidth, _ROW_HEIGHT)))
        FunctionalBase.__init__(self)
        self._rowWidth = rowWidth
        self._label = PlainText(UI.DefaultFont, labelText, _LABEL_FONT_SIZE)
        labelBounds = self._label.getLocalBounds()
        labelY = (float(_ROW_HEIGHT) - labelBounds.size.y) / 2.0
        self._label.setPosition(Vector2f(float(_LABEL_X) - labelBounds.position.x, labelY))
        self._label.setOrigin(Vector2f(0.0, 0.0))
        self.addChild(self._label)
        self._checkBox = CheckBox(checked, checkBoxSize, windowSkin, onCheckedChanged)
        checkBoxY = (float(_ROW_HEIGHT) - float(checkBoxSize)) / 2.0
        self._checkBox.getField().setPosition(Vector2f(float(rowWidth) / 2.0, checkBoxY))
        self.addChild(self._checkBox.getField())
        self.addConfirmCallback(self._onConfirmToggle)
        self.setOrigin(Vector2f(0.0, 0.0))

    def getCheckBox(self) -> CheckBox:
        r"""\brief Get the row CheckBox coordinator.

        - \return  Nested CheckBox instance
        """
        return self._checkBox

    def getChildren(self) -> list:
        r"""\brief Disable default canvas child traversal; this row renders as a unit.

        - \return  Empty list
        """
        return []

    def getSize(self) -> Vector2f:
        r"""\brief Get the row size for ListView layout.

        - \return  Row width and height in logical UI units
        """
        return Vector2f(float(self._rowWidth), float(_ROW_HEIGHT))

    def getLocalBounds(self) -> FloatRect:
        r"""\brief Get the visible row bounds matching the checkbox row height.

        - \return  Local bounds for hit-testing and pointer focus
        """
        return FloatRect(Vector2f(0.0, 0.0), self.getSize())

    def onTick(self, deltaTime: float) -> None:
        r"""\brief Update nested widgets and render the row canvas.

        - \param deltaTime  Elapsed time in seconds
        """
        self._buildRenderQueue()
        self.render()

    def _onConfirmToggle(self, kwargs: dict) -> None:
        self._checkBox.toggle()

File: Sample/Scripts/Source/UI/Parts/ConfigWindow/ConfigCheckBoxRow.lua
local ConfigRowControllerBase = require("Source.UI.Parts.ConfigWindow.ConfigRow")
local Ui = require("Source.UI.Ui")

local _ROW_HEIGHT = 32

local ConfigCheckBoxRowUI = {}

function ConfigCheckBoxRowUI:init(labelText, rowWidth, checkBoxSize, windowSkin, checked, onCheckedChanged)
    self._labelText = labelText
    self._checkBoxSize = checkBoxSize
    self._windowSkin = windowSkin
    if checked == nil then
        checked = false
    end
    self._checked = checked
    self._onCheckedChanged = onCheckedChanged
    super(ConfigCheckBoxRowUI, self).init(nil, rowWidth, _ROW_HEIGHT)
end

function ConfigCheckBoxRowUI:bind()
    self._label = self:requireControl("Label")
    self._checkBox = self:requireControl("CheckBox")
    self._checkBox:setCanReceiveFocus(false)
    self._checkBox:resize(sf.Vector2f.new(self._checkBoxSize, self._checkBoxSize))
    self._checkBox:setWindowSkin(self._windowSkin)
    self._checkBox:setChecked(self._checked)
    self._checkBox:setOnCheckedChanged(self._onCheckedChanged)
    self.root:addConfirmCallback(function (_, kwargs)
        self:_onConfirmToggle(kwargs)
    end)
end

function ConfigCheckBoxRowUI:refresh()
    self:setText("Label", self._labelText)
end

function ConfigCheckBoxRowUI:getCheckBox()
    return self._checkBox
end

function ConfigCheckBoxRowUI:setActive(active)
    super(ConfigCheckBoxRowUI, self).setActive(active)
    self._checkBox:setActive(active)
end

function ConfigCheckBoxRowUI:dispose()
    if self._checkBox ~= nil then
        self._checkBox:setOnCheckedChanged(nil)
    end
    self.root:addConfirmCallback(nil)
    self._onCheckedChanged = nil
    self._checkBox = nil
    super(ConfigCheckBoxRowUI, self).dispose()
end

---@param _kwargs table
function ConfigCheckBoxRowUI:_onConfirmToggle(_kwargs)
    self._checkBox:toggle()
end

return Ui.Define("Parts/ConfigWindow/ConfigCheckBoxRow", ConfigCheckBoxRowUI, ConfigRowControllerBase)
